backend/plugin/mfa/otp/otp.go

```diff
@@ -0,0 +1,102 @@
+// Package otp is the plugin for OTP Multi-Factor Authentication.
+package otp
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pquerna/otp/totp"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	// issuerName is the name of the issuer of the OTP token.
+	issuerName = "Bytebase"
+	// secondsInMinute is the number of seconds in a minute.
+	secondsInMinute = int64(time.Minute.Seconds())
+	// maxPastSecretCount is the maximum number of past secret we will check.
+	maxPastSecretCount = 5
+)
+
+// removeSecondsFromTimestamp removes the seconds from the timestamp.
+func removeSecondsFromTimestamp(timestamp time.Time) int64 {
+	return timestamp.Unix() - timestamp.Unix()%secondsInMinute
+}
+
+// TimeBasedReader is a reader that returns the same value for the same timestamp.
+type TimeBasedReader struct {
+	reader *strings.Reader
+}
+
+// NewTimeBasedReader creates a new TimeBasedReader with the given timestamp.
+func NewTimeBasedReader(timestamp time.Time) *TimeBasedReader {
+	return &TimeBasedReader{
+		reader: strings.NewReader(strconv.FormatInt(removeSecondsFromTimestamp(timestamp), 10)),
+	}
+}
+
+func (r *TimeBasedReader) Read(p []byte) (int, error) {
+	return r.reader.Read(p)
+}
+
+// GenerateSecret generates a new secret for the given account name and timestamp.
+func GenerateSecret(accountName string, timestamp time.Time) (string, error) {
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      issuerName,
+		AccountName: accountName,
+		Rand:        NewTimeBasedReader(timestamp),
+	})
+	if err != nil {
+		return "", err
+	}
+	return key.Secret(), nil
+}
+
+// GetPastSecrets returns the past 5 secrets for the given account name and timestamp.
```

Comment:
GetPastSecrets -> GeValidSecrets

Also do not need to include 5, otherwise, it's easy to forget to update this if we change maxPastSecretCount

Reply:
Updated